### PR TITLE
Porting TextBlock, Inlines, Flow Layout from WPF

### PR DESCRIPTION
**Pull request by shartte**
As requested by @Gillibald, a branch for discussing work on porting the inline work from WPF.

### CODE REVIEW

File: src/Avalonia.Documents/Documents/IAddChild.cs [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+namespace System.Windows.Markup 
+{
+    ///<summary>
+    /// The IAddChild interface is used for parsing objects that
+    /// allow objects or text underneath their tags in markup that
+    /// do not map directly to a property. 
+    ///</summary>
+    //[Obsolete("IAddChild is obsolete, use the [ContentProperty] attribute instead")]
+    public interface IAddChild
```
Comment (Gillibald) [inline src/Avalonia.Documents/Documents/IAddChild.cs:26]: This interface is redundant. Avalonia's compiler is looking for an `Add` method by default. So every class that previously implemented this interface should expose `void Add(object child)`

File: src/Avalonia.Documents/Documents/InlineCollection.cs [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+            AddText(text, false /* returnIndex */);
+        }
+
+        /// <summary>
+        /// Adds an implicit InlineUIContainer with a given UIElement in it.
+        /// </summary>
+        /// <param name="uiElement">
+        /// UIElement set as a Child property for the implicit InlineUIContainer.
+        /// </param>
+        public void Add(IControl uiElement)
```
Comment (Gillibald) [inline src/Avalonia.Documents/Documents/InlineCollection.cs:109]: Previously `IAddChild`

File: src/Avalonia.Documents/Documents/Internal/NewTextBlock.cs [hunk 3]
```diff
@@ -0,0 +1,3311 @@
...
+        public static readonly StyledProperty<Thickness> PaddingProperty =
+            Decorator.PaddingProperty.AddOwner<NewTextBlock>();
+
+        // TODO: Define these attached properties elsewhere (e.g. on a Text class) and AddOwner
+        // them into NewTextBlock.
+
+        /// <summary>
+        /// Defines the <see cref="FontFamily"/> property.
+        /// </summary>
+        public static readonly AttachedProperty<FontFamily> FontFamilyProperty =
```
Comment (Gillibald) [inline src/Avalonia.Documents/Documents/Internal/NewTextBlock.cs:68]: These are defined on TextElement so we should inherit them from there

File: src/Avalonia.Documents/Documents/Internal/IdentityTransform.cs [hunk 4]
```diff
@@ -0,0 +1,15 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using Avalonia.Media;
+
+namespace Avalonia.Documents.Internal
+{
+    public class IdentityTransform : Transform
+    {
+        // TODO: This replaces Transform.Identity. Is this okay, or should this be moved into Avalonia?
```
Comment (Gillibald) [inline src/Avalonia.Documents/Documents/Internal/IdentityTransform.cs:10]: Transform.Identity should be used I guess.

File: src/Avalonia.Visuals/Media/TextFormatting/TextFormatterImpl.cs [hunk 5]
```diff
@@ -548,5 +602,13 @@ public bool MoveNext()
                 return !(Current is TextEndOfLine);
             }
         }
+
+        private static LineBreakEnumerator GetLineBreakEnumerator(TextRun run)
```
Comment (Gillibald) [inline src/Avalonia.Visuals/Media/TextFormatting/TextFormatterImpl.cs:606]: What is the reason behind this change?

File: src/Avalonia.Visuals/Media/TextFormatting/TextFormatterImpl.cs [hunk 6]
```diff
@@ -548,5 +602,13 @@ public bool MoveNext()
                 return !(Current is TextEndOfLine);
```
Comment (Gillibald) [inline src/Avalonia.Visuals/Media/TextFormatting/TextFormatterImpl.cs:602]: `return true`

File: src/Avalonia.Visuals/Media/TextFormatting/TextFormatterImpl.cs [hunk 7]
```diff
@@ -311,6 +327,11 @@ internal static SplitTextRunsResult SplitTextRuns(List<ShapedTextCharacters> tex
                 currentLength += textRun.Text.Length;
             }
 
+            if (textRunEnumerator.Current is TextEndOfParagraph)
```
Comment (Gillibald) [inline src/Avalonia.Visuals/Media/TextFormatting/TextFormatterImpl.cs:330]: Remove this and introduce a default case in the switch that just adds the remaining run types to the list

File: src/Avalonia.Visuals/Media/TextFormatting/TextFormatterImpl.cs (hunk 6 above)
Comment (Gillibald) [reply src/Avalonia.Visuals/Media/TextFormatting/TextFormatterImpl.cs:602]: After this change, every `ITextSource` that previously used `EndOfLine` as the end marker should now return null

File: src/Avalonia.Visuals/Media/TextFormatting/TextFormatterImpl.cs (hunk 7 above)
Comment (Gillibald) [reply src/Avalonia.Visuals/Media/TextFormatting/TextFormatterImpl.cs:330]:
```
case TextEndOfLine endOfLine:
                    {
                        textRuns.Add(endOfLine);
                        break;
                    }
```

File: src/Avalonia.Documents/Documents/InlineCollection.cs (hunk 2 above)
Comment (shartte) [reply src/Avalonia.Documents/Documents/InlineCollection.cs:109]: I changed the param to IControl (for now) here because TextElement's extend from StyledElement as far as I know.
This method however is meant for adding elements that should be auto-wrapped, such as images, etc. I think IControl might be a good target interface for that.

**Comment by Gillibald** [comment]
I just realized that I have to port https://docs.microsoft.com/en-us/dotnet/api/system.windows.media.textformatting.textline.length?view=net-5.0 otherwise we can't port some parts

**Comment by Gillibald** [comment]
This probably involves removing `TextRange` from `TextLine`
This also might involve changing how cluster values are used. So every line has cluster values starting at zero.

**Comment by Gillibald** [comment]
There is something about those synthetic characters. When I add a LineBreak to the TextBlock it always uses TextSourceLength=2 for the LineBreak. I think we need to treat LineBreak etc. as actual text.

**Comment by Gillibald** [comment]
This line in particular https://github.com/dotnet/wpf/blob/d49f8ddb889b5717437d03caa04d7c56819c16aa/src/Microsoft.DotNet.Wpf/src/PresentationFramework/MS/Internal/Text/ComplexLine.cs#L395

**Comment by shartte** [comment]
Yes, that is extremely weird. In this branch, my current hack was to actually define `Text` for EndOfLine/EndOfParagraph and make it contain actual text (`\n` and the paragraph separator unicode character, respectively).
However. Is my understanding correct that the text source positions are completely separated from the text that actually gets shaped in the end? Hence the differentiation between `TextSourceLength` and actual Length? We currently wildly mix the two concepts.

**Comment by Gillibald** [comment]
TextSourceLength is just meant for querying runs from the text source

**Comment by Gillibald** [comment]
I think WPF treats `TextEndOfLine` as part of the TextLine and therefore the TextSourceLength contributes to the final length of the line. This is why they subtract the synthetic character count from `TextLine.Length`

**Comment by Gillibald** [comment]
We probably want to fire up a WPF application and test `TextFormatter.FormatLine` in combination with `TextEndOfLine` to find out the expected results.

**Comment by Gillibald** [comment]
I have run something under WPF. So the full length of a `TextEndOfLine` run adds to the total length of the `TextLine`. `TextEndOfLine` is always added to the list of runs.
Screenshot 2021-02-11 135223
Screenshot 2021-02-11 135239

**Comment by Gillibald** [comment]
This can be used to investigate WPF's behavior
[WpfApp1.zip](https://github.com/AvaloniaUI/Avalonia/files/5965356/WpfApp1.zip)

**Comment by Gillibald** [comment]
I think we don't have to process a synthetic LineBreak like WPF does. We can just introduce a flat `EOT` and use that instead. Adding TextEndOfLine to the list of runs just makes the code more complicated but isn't any useful in the end.

**Comment by shartte** [comment]
@Gillibald Do you mean EndOfParagraph? I think the hickup in TextBlock was not about end of line, rather EndOfParagraph.
It's hard to judge at this point, how this will interact with the FlowDocument later, since TextBlock only displays a single paragraph at a time, apparently. Although I also thought that potentially just indicating EndOfText/EndOfLine via a boolean Property might be sufficient.

**Comment by Gillibald** [comment]
TextEndOfParagraph is a TextEndOfLine. For TextBlock `LineBreak` is transformed into `TextEndOfLine` on top of that when the TextSource reaches its end a TextEndOfParagraph is produced. Both are processed in the same way by the standard TextFormatter. FlowDocument most likely processes TextEndOfParagraph differently.

**Comment by shartte** [comment]
I am talking specifically about this: https://github.com/dotnet/wpf/blob/master/src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/Controls/TextBlock.cs#L1406

**Comment by Gillibald** [comment]
Y that implicitly looks at the last run in the line. If it is a TextEndOfParagraph this is true. I don't think that synthetic runs should contribute to the text metrics.

**Comment by shartte** [comment]
I think the problem was that it looks at the `trailing whitespace` field as a performance optimization to quickly test if there's any trailing whitespace (which includes EndOfLine runs apparently) in the line. And that field comes from the metrics.

It feels convoluted.

**Comment by Gillibald** [comment]
Y we don't need to mimic that. Just hold a flag in the text line metric that indicates synthetic characters (TextEndOfLine). Adding synthetic characters to the NewLineLength is a hack.

**Comment by Gillibald** [comment]
We can close it. I need to rework most of it.

**Comment by PeteMichaud** [comment]
Should I interpret this closure as meaning that AvaloniaEdit does not currently support rich text?

**Comment by Gillibald** [comment]
This work was about porting WPF's text tree and in some extend bringing FlowDocument to Avalonia. It was an attempt of multiple and was closed because it wasn't in a state we could move further.

Avalonia now supports proper whitespace handling via XamlX and also has inlines support.

We will most likely port WPF's text tree and other parts of FlowDocument in the near future. This is highly reliant on demand and funding.

**Comment by mlsomers** [comment]
Sorry for being off-topic but I'd just like to throw this idea in here.

It's going to be a huge task but if you ever get to it, please consider to keep an option open for some features that were lacking in the original WPF and finally rendered it unusable for my purpose:

`<FlowDocument Language="nl" IsHyphenationEnabled="True">`

[IsHyphenationEnabled](https://docs.microsoft.com/en-us/dotnet/api/system.windows.documents.flowdocument.ishyphenationenabled) was restricted to only four languages: English, Spanish, French, and German. It is related to the spell-checking dictionaries which could be [extended for extra jargon](https://web.archive.org/web/20160309181414/http://blogs.msdn.com/b/text/archive/2009/10/02/custom-dictionaries.aspx) but did not allow adding any full dictionary.

If this feature would have been available at that time it would have entirely removed Adobe InDesign from my workflow.
The text could be rendered to XPS using a DocumentPaginator, from there to PDF.

It's probably far out of scope, but if you could ever get Avalonia to natively render PDF, your audience would instantly increase. This is a feature that some Winforms and WPF users have wanted for decades. There are many commercial solutions (most of them not native .Net). The best open source PDF in .Net that I am aware of is [PdfSharp](http://www.pdfsharp.net/MainPage.ashx) which is far from mature.

Bottom line: If you could bring typesetting and native PDF into scope I think there would be allot more demand and possibly funding from a wider audience outside the cross-platform-ui domain...
